libbpf-tools/vfsstat.c

```diff
@@ -0,0 +1,210 @@
+// Copyright (c) 2020 Anton Protopopov
+//
+// Based on vfsstat(8) from BCC by Brendan Gregg
+#include <sys/resource.h>
+#include <argp.h>
+#include <unistd.h>
+#include <time.h>
+#include <bpf/bpf.h>
+#include "vfsstat.h"
+#include "vfsstat.skel.h"
+
+const char *argp_program_version = "vfsstat 0.1";
+const char *argp_program_bug_address = "<bpf@vger.kernel.org>";
+static const char argp_program_doc[] =
+	"\nvfsstat: Count some VFS calls\n"
+	"\n"
+	"EXAMPLES:\n"
+	"    vfsstat      # interval one second\n"
+	"    vfsstat 5 3  # interval five seconds, three output lines\n";
+static char args_doc[] = "[interval [count]]";
+
+static const struct argp_option opts[] = {
+	{ "verbose", 'v', NULL, 0, "Verbose debug output" },
+	{},
+};
+
+static struct env {
+	bool verbose;
+	int count;
+	int interval;
+} env = {
+	.interval = 1,	/* once a second */
+};
+
+static error_t parse_arg(int key, char *arg, struct argp_state *state)
+{
+	long interval;
+	long count;
+
+	switch (key) {
+	case 'v':
+		env.verbose = true;
+		break;
+	case ARGP_KEY_ARG:
+		switch (state->arg_num) {
+		case 0:
+			interval = strtol(arg, NULL, 10);
+			if (errno || interval <= 0 || interval > INT_MAX) {
+				fprintf(stderr, "invalid interval: %s\n", arg);
+				argp_usage(state);
+			}
+			env.interval = interval;
+			break;
+		case 1:
+			count = strtol(arg, NULL, 10);
+			if (errno || count < 0 || count > INT_MAX) {
+				fprintf(stderr, "invalid count: %s\n", arg);
+				argp_usage(state);
+			}
+			env.count = count;
+			break;
+		default:
+			argp_usage(state);
+			break;
+		}
+		break;
+	default:
+		return ARGP_ERR_UNKNOWN;
+	}
+	return 0;
+}
+
+static int libbpf_print_fn(enum libbpf_print_level level,
+		const char *format, va_list args)
+{
+	if (level == LIBBPF_DEBUG && !env.verbose)
+		return 0;
+	return vfprintf(stderr, format, args);
+}
+
+static int bump_memlock_rlimit(void)
+{
+	struct rlimit rlim_new = {
+		.rlim_cur	= RLIM_INFINITY,
+		.rlim_max	= RLIM_INFINITY,
+	};
+
+	return setrlimit(RLIMIT_MEMLOCK, &rlim_new);
+}
+
+static const char *Strftime(char *s, size_t max, const char *format)
+{
+	struct tm *tm;
+	time_t t;
+
+	t = time(NULL);
+	tm = localtime(&t);
+	if (tm == NULL) {
+		fprintf(stderr, "localtime: %s\n", strerror(errno));
+		return "<failed>";
+	}
+	if (strftime(s, max, format, tm) == 0) {
+		fprintf(stderr, "strftime error\n");
+		return "<failed>";
+	}
+	return s;
+}
+
+static const char *stat_types_names[] = {
+	[S_READ] = "READ",
+	[S_WRITE] = "WRITE",
+	[S_FSYNC] = "FSYNC",
+	[S_OPEN] = "OPEN",
+	[S_CREATE] = "CREATE",
+};
+
+static void print_header(void)
+{
+	printf("%-8s  ", "TIME");
+	for (int i = 0; i < S_MAXSTAT; i++)
+		printf(" %6s/s", stat_types_names[i]);
+	printf("\n");
+}
+
+static void print_and_reset_stats(int fd)
+{
+	__u32 count = S_MAXSTAT;
+	__u64 vals[S_MAXSTAT];
+	int keys[S_MAXSTAT];
+	void *batch;
+	char s[16];
+	int err;
+
+	err = bpf_map_lookup_batch(fd, NULL, &batch, keys, vals, &count, NULL);
```

Comment:
batch is very recent APIs, this will limit which kernel versions are going to be supported. With using normal array on BPF side, you don't need batching, and your code actually going to be faster due to memory-mapped array in user-space.